### PR TITLE
feat: Overflow menu should be registered in overflowManager

### LINKED ISSUE #25090: [Bug]: Overflow menu does not get replaced with last invisible element
### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
System:
    OS: Linux 5.15 Ubuntu 20.04.3 LTS (Focal Fossa)
    CPU: (8) x64 Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz
    Memory: 11.44 GB / 15.42 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash
  Browsers:
    Chrome: 104.0.5112.101
```

### Are you reporting Accessibility issue?

no

### Reproduction

https://codesandbox.io/s/unruffled-einstein-i39odt?file=/example.tsx:1869-1876

## Actual Behavior
When expanding the overflow container, sometimes the overflow menu can take up space that would show the last invisible element anyway

## Expected Behavior
If the space occupied by the overflow menu can fit the last invisible element, it should do so

https://user-images.githubusercontent.com/20744592/194061102-2f1af030-2323-4826-be41-6d1312f1d983.mov

### Logs

_No response_

### Requested priority

High

### Products/sites affected

_No response_

### Are you willing to submit a PR to fix?

yes

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

### PR DESCRIPTION
**Pull request by ling1726**
## Current Behavior

The overflow manager treats the overflow menu like any other overflow item. This causes the unfortunate bug that there is no way to whether the last invisible item can simply replace the overflow menu.
## New Behavior

Adds special treatment to the overflow menu so it is a special kind of item in the overflow manager.

Updates `processOverflow` so that there is a special case at the end to check for the case that the overflow menu can be replaced with the last overflow item.

## Related Issue(s)

Fixes #25090

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports                                                                                                        |    Baseline (minified/GZIP) |                          PR |                                                                                                                                                                                                  Change |
| :----------------------------------------------------------------------------------------------------------------------- | --------------------------: | --------------------------: | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| <samp>priority-overflow</samp> <br /> <abbr title='bundle-size/overflow-manager.fixture.js'>createOverflowManager</abbr> |  `2.936 kB`<br />`1.212 kB` |  `3.088 kB`<br />`1.273 kB` | `152 B` <br />`61 B`  |
| <samp>react-overflow</samp> <br /> <abbr title='bundle-size/hooks.fixture.js'>hooks only</abbr>                          | `10.685 kB`<br />`4.104 kB` | `10.982 kB`<br />`4.185 kB` | `297 B` <br />`81 B`  |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                             |          Size (minified/GZIP) |
| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `188.815 kB`<br />`52.287 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>                            |    `33.4 kB`<br />`11.008 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr>                                                         |    `5.857 kB`<br />`1.978 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/443c503b94e602e55857e4e311413c6257d60766'>443c503b94e602e55857e4e311413c6257d60766</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/Avatar-mount_base.html">1270</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/Avatar-mount.html">1283</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/Button-mount_base.html">918</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/Button-mount.html">932</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/FluentProvider-mount_base.html">1466</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/FluentProvider-mount.html">1455</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/FluentProviderWithTheme-mount_base.html">571</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/FluentProviderWithTheme-mount.html">576</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">538</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">536</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">578</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">572</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/MakeStyles-mount_base.html">1917</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/MakeStyles-mount.html">1939</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/SpinButton-mount_base.html">2314</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25091/perf-test-react-components/SpinButton-mount.html">2322</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: packages/react-components/priority-overflow/src/priorityQueue.ts [hunk 1]
```diff
@@ -88,13 +88,14 @@ export function createPriorityQueue<T>(compare: PriorityQueueCompareFn<T>): Prio
...
-    return arr.indexOf(item) >= 0;
+    const index = arr.indexOf(item);
+    return index >= 0 && index < size;
   };
 
   const remove = (item: T) => {
     const i = arr.indexOf(item);
 
-    if (i === -1) {
+    if (i === -1 || i >= size) {
```
Comment (ling1726) [inline packages/react-components/priority-overflow/src/priorityQueue.ts:98]: these were absolutely awful bugs

File: packages/react-components/react-overflow/src/useOverflowMenu.ts [hunk 2]
```diff
@@ -2,32 +2,27 @@ import * as React from 'react';
...
-        deregisterItem();
-        element?.removeAttribute(DATA_OVERFLOW_MENU);
-      };
+    if (!element || !isOverflowing) {
+      offsetRef.current = 0;
+      return;
     }
-  }, [registerItem, isOverflowing, elementId]);
+
+    offsetRef.current = 1;
```
Comment (layershifter) [inline packages/react-components/react-overflow/src/useOverflowMenu.ts:22]: What is `offsetRef`? I don't see any usage/assignments outside of the effect.

File: packages/react-components/priority-overflow/src/priorityQueue.ts (hunk 1 above)
Comment (layershifter) [inline packages/react-components/priority-overflow/src/priorityQueue.ts:98]: How `i` could be bigger than `size`?

File: packages/react-components/priority-overflow/src/overflowManager.ts [hunk 3]
```diff
@@ -121,6 +122,8 @@ export function createOverflowManager(): OverflowManager {
       return;
     }
 
+    const overflowMenuSize = overflowMenu ? getOffsetSize(overflowMenu) : 0;
```
Comment (layershifter) [inline packages/react-components/priority-overflow/src/overflowManager.ts:125]: nit: I would `overflowMenuSize` to reflect that it's offset

File: packages/react-components/react-overflow/src/useOverflowMenu.ts (hunk 2 above)
Comment (ling1726) [reply packages/react-components/react-overflow/src/useOverflowMenu.ts:22]: it's cruft :)

File: packages/react-components/priority-overflow/src/priorityQueue.ts (hunk 1 above)
Comment (ling1726) [reply packages/react-components/priority-overflow/src/priorityQueue.ts:98]: there is no array 'deletion' happening here, removed items simply get pushed out of bounds

**Review by layershifter — APPROVED** [review]

**Pull request merged**